k2/torch/python/csrc/fsa.cu

```diff
@@ -0,0 +1,153 @@
+/**
+ * @brief python wrapper for Ragged<Arc>
+ *
+ * @copyright
+ * Copyright      2021  Xiaomi Corp.  (authors: Fangjun Kuang,
+ *                                              Wei Kang)
+ *
+ * @copyright
+ * See LICENSE for clarification regarding multiple authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "k2/csrc/ragged.h"
+#include "k2/torch/csrc/fsa_class.h"
+#include "k2/torch/csrc/torch_utils.h"
+#include "k2/torch/python/csrc/fsa.h"
+#include "k2/torch/python/csrc/pybind_utils.h"
+
+namespace k2 {
+
+void PybindFsaClass(py::module &m) {
+  // TODO(fangjun): Add more doc to doc/fsa.cu for k2.ragged.FSA
+  py::class_<FsaClass> fsa(m, "Fsa");
+
+  fsa.def(py::init([](const std::string &s,
+                      torch::optional<std::vector<std::string>>
+                          extra_label_names = {}) -> std::unique_ptr<FsaClass> {
+            return std::make_unique<FsaClass>(
+                s, extra_label_names.value_or(std::vector<std::string>{}));
+          }),
+          py::arg("s"), py::arg("extra_label_names") = py::none());
+
+  fsa.def("__str__", &FsaClass::ToString);
+  fsa.def("__repr__", &FsaClass::ToString);
+
+  fsa.def("requires_grad_", &FsaClass::SetRequiresGrad,
+          py::arg("requires_grad") = true);
+
+  fsa.def(
+      "to",
+      static_cast<FsaClass (FsaClass::*)(torch::Device) const>(&FsaClass::To),
+      py::arg("device"));
+
+  fsa.def("to",
+          static_cast<FsaClass (FsaClass::*)(const std::string &) const>(
+              &FsaClass::To),
+          py::arg("device"));
+
+  fsa.def(
+      "set_scores_stochastic_",
+      [](FsaClass &self, torch::Tensor scores) -> void {
+        self.SetScoresStochastic(scores);
+      },
+      py::arg("scores"));
+
+  fsa.def("arc_sort", &FsaClass::ArcSort);
+
+  fsa.def(
+      "__setattr__",
+      [](FsaClass &self, const std::string &name, py::object value) -> void {
+        if (name == "scores") {
+          self.SetScores(value.cast<torch::Tensor>());
+        } else {
+          self.SetAttr(name, ToIValue(value));
+        }
+      });
+
+  fsa.def("__getattr__",
+          [](FsaClass &self, const std::string &name) -> py::object {
+            torch::IValue ivalue;
+            try {
+              ivalue = self.GetAttr(name);
```

Comment:
Can you first use `HasAttr` to check if the attr exists?
It can avoid throwing an exception in C++.